File: 2025/12/03/evaluating-llm-systems-production/python/src/logger.py
"""
Logging module for LLM calls.
Captures input, output, metadata, and performance metrics.
"""

import json
import hashlib
import re
from datetime import datetime
from typing import Dict, List, Optional, Any
from dataclasses import dataclass, asdict


def redact_pii(text: str) -> str:
    """Redact PII from text."""
    # Email addresses
    text = re.sub(r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b', '[EMAIL]', text)
    # Phone numbers
    text = re.sub(r'\b\d{3}-\d{3}-\d{4}\b', '[PHONE]', text)
    # Credit cards
    text = re.sub(r'\b\d{4}[\s-]?\d{4}[\s-]?\d{4}[\s-]?\d{4}\b', '[CARD]', text)
    return text


def hash_user_id(user_id: str) -> str:
    """Hash user ID for privacy."""
    return hashlib.sha256(user_id.encode()).hexdigest()[:16]


@dataclass
class LLMLog:
    """Log record for a single LLM call."""
    request_id: str
    timestamp: str
    user_id_hash: str
    session_id: Optional[str] = None
    input: Optional[Dict[str, Any]] = None
    output: Optional[Dict[str, Any]] = None
    model: Optional[Dict[str, Any]] = None
    prompt: Optional[Dict[str, Any]] = None
    performance: Optional[Dict[str, Any]] = None
    experiment: Optional[Dict[str, Any]] = None
    feedback: Optional[Dict[str, Any]] = None

    def to_dict(self) -> Dict:
        """Convert to dictionary."""
        return asdict(self)

    def to_json(self) -> str:
        """Convert to JSON string."""
        return json.dumps(self.to_dict(), indent=2)


class LLMLogger:
    """Logger for LLM calls."""
    
    def __init__(self, log_file: Optional[str] = None, redact_pii_enabled: bool = True):
        self.log_file = log_file
        self.redact_pii_enabled = redact_pii_enabled
        self.logs: List[LLMLog] = []
    
    def log_call(
        self,
        request_id: str,
        user_id: str,
        query: str,
        response: str,
        model_name: str,
        model_version: str,
        prompt_version: str,
        latency_ms: int,
        tokens_used: int,
        cost_usd: float,
        session_id: Optional[str] = None,
        context: Optional[List[str]] = None,
        experiment_variant: Optional[str] = None,
        experiment_cohort: Optional[str] = None,
        temperature: float = 0.7
    ) -> LLMLog:
        """Log an LLM call."""
        
        # Redact PII if enabled
        query_redacted = redact_pii(query) if self.redact_pii_enabled else query
        response_redacted = redact_pii(response) if self.redact_pii_enabled else response
        
        # Hash user ID
        user_id_hash = hash_user_id(user_id)
        
        # Create log record
        log = LLMLog(
            request_id=request_id,
            timestamp=datetime.utcnow().isoformat(),
            user_id_hash=user_id_hash,
            session_id=session_id,
            input={
                "query": query_redacted,
                "context": context or []
            },
            output={
                "text": response_redacted,
                "tokens_used": tokens_used
            },
            model={
                "name": model_name,
                "version": model_version,
                "temperature": temperature
            },
            prompt={
                "template_version": prompt_version,
                "template_hash": hashlib.md5(prompt_version.encode()).hexdigest()[:8]
            },
            performance={
                "latency_ms": latency_ms,
                "cost_usd": cost_usd
            },
            experiment={
                "variant": experiment_variant or "baseline",
                "cohort": experiment_cohort or "control"
            } if experiment_variant else None
        )
        
        # Store log
        self.logs.append(log)
        
        # Write to file if specified
        if self.log_file:
            with open(self.log_file, 'a') as f:
                f.write(log.to_json() + '\n')
        
        return log
    
    def add_feedback(
        self,
        request_id: str,
        thumbs_up: Optional[bool] = None,
        rating: Optional[int] = None,
        labels: Optional[Dict[str, str]] = None
    ):
        """Add feedback to an existing log."""
        for log in self.logs:
            if log.request_id == request_id:
                log.feedback = {
                    "thumbs_up": thumbs_up,
                    "rating": rating,
                    "labels": labels or {},
                    "timestamp": datetime.utcnow().isoformat()
                }
                break
    
    def get_logs(self, user_id_hash: Optional[str] = None, variant: Optional[str] = None) -> List[LLMLog]:
        """Get logs, optionally filtered."""
        logs = self.logs
        if user_id_hash:
            logs = [l for l in logs if l.user_id_hash == user_id_hash]
        if variant:
            logs = [l for l in logs if l.experiment and l.experiment.get("variant") == variant]
        return logs

File: 2025/12/03/evaluating-llm-systems-production/typescript/src/logger.ts
/**
 * Logging module for LLM calls.
 * Captures input, output, metadata, and performance metrics.
 */

import * as crypto from 'crypto';

export interface LLMLog {
  request_id: string;
  timestamp: string;
  user_id_hash: string;
  session_id?: string;
  input?: {
    query: string;
    context?: string[];
  };
  output?: {
    text: string;
    tokens_used?: number;
  };
  model?: {
    name: string;
    version: string;
    temperature?: number;
  };
  prompt?: {
    template_version: string;
    template_hash: string;
  };
  performance?: {
    latency_ms: number;
    cost_usd: number;
  };
  experiment?: {
    variant: string;
    cohort: string;
  };
  feedback?: {
    thumbs_up?: boolean;
    rating?: number;
    labels?: Record<string, string>;
    timestamp: string;
  };
}

function redactPII(text: string): string {
  // Email addresses
  text = text.replace(/\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/g, '[EMAIL]');
  // Phone numbers
  text = text.replace(/\b\d{3}-\d{3}-\d{4}\b/g, '[PHONE]');
  // Credit cards
  text = text.replace(/\b\d{4}[\s-]?\d{4}[\s-]?\d{4}[\s-]?\d{4}\b/g, '[CARD]');
  return text;
}

function hashUserId(userId: string): string {
  return crypto.createHash('sha256').update(userId).digest('hex').substring(0, 16);
}

export class LLMLogger {
  private logs: LLMLog[] = [];
  private logFile?: string;
  private redactPIIEnabled: boolean;

  constructor(logFile?: string, redactPIIEnabled: boolean = true) {
    this.logFile = logFile;
    this.redactPIIEnabled = redactPIIEnabled;
  }

  logCall(params: {
    request_id: string;
    user_id: string;
    query: string;
    response: string;
    model_name: string;
    model_version: string;
    prompt_version: string;
    latency_ms: number;
    tokens_used: number;
    cost_usd: number;
    session_id?: string;
    context?: string[];
    experiment_variant?: string;
    experiment_cohort?: string;
    temperature?: number;
  }): LLMLog {
    const {
      request_id,
      user_id,
      query,
      response,
      model_name,
      model_version,
      prompt_version,
      latency_ms,
      tokens_used,
      cost_usd,
      session_id,
      context,
      experiment_variant,
      experiment_cohort,
      temperature = 0.7,
    } = params;

    // Redact PII if enabled
    const queryRedacted = this.redactPIIEnabled ? redactPII(query) : query;
    const responseRedacted = this.redactPIIEnabled ? redactPII(response) : response;

    // Hash user ID
    const userIdHash = hashUserId(user_id);

    // Create log record
    const log: LLMLog = {
      request_id,
      timestamp: new Date().toISOString(),
      user_id_hash: userIdHash,
      session_id,
      input: {
        query: queryRedacted,
        context: context || [],
      },
      output: {
        text: responseRedacted,
        tokens_used,
      },
      model: {
        name: model_name,
        version: model_version,
        temperature,
      },
      prompt: {
        template_version: prompt_version,
        template_hash: crypto
          .createHash('md5')
          .update(prompt_version)
          .digest('hex')
          .substring(0, 8),
      },
      performance: {
        latency_ms,
        cost_usd,
      },
      experiment:
        experiment_variant
          ? {
              variant: experiment_variant || 'baseline',
              cohort: experiment_cohort || 'control',
            }
          : undefined,
    };

    // Store log
    this.logs.push(log);

    // Write to file if specified (in Node.js environment)
    if (this.logFile && typeof require !== 'undefined') {
      const fs = require('fs');
      fs.appendFileSync(this.logFile, JSON.stringify(log) + '\n');
    }

    return log;
  }

  addFeedback(
    request_id: string,
    feedback: {
      thumbs_up?: boolean;
      rating?: number;
      labels?: Record<string, string>;
    }
  ): void {
    for (const log of this.logs) {
      if (log.request_id === request_id) {
        log.feedback = {
          ...feedback,
          timestamp: new Date().toISOString(),
        };
        break;
      }
    }
  }

  getLogs(filters?: {
    user_id_hash?: string;
    variant?: string;
  }): LLMLog[] {
    let logs = this.logs;
    if (filters?.user_id_hash) {
      logs = logs.filter((l) => l.user_id_hash === filters.user_id_hash);
    }
    if (filters?.variant) {
      logs = logs.filter(
        (l) => l.experiment?.variant === filters.variant
      );
    }
    return logs;
  }
}

